apps/emqx_mt/src/emqx_mt_state.erl

```diff
@@ -0,0 +1,287 @@
+    evict_ccache/1,
+    list_clients/3,
+    is_known_client/2
+]).
+
+-export([
+    add/3,
+    del/1,
+    clear_for_node/1,
+    clear_self_node/0
+]).
+
+-include("emqx_mt.hrl").
+-include_lib("stdlib/include/ms_transform.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+-define(EMQX_MT_SHARD, emqx_mt_shard).
+
+%% mria tables
+-define(RECORD_TAB, emqx_mt_record).
+-define(COUNTER_TAB, emqx_mt_counter).
+-define(NS_TAB, emqx_mt_ns).
+
+%% ets tables
+-define(MONITOR_TAB, emqx_mt_monitor).
+-define(CCACHE_TAB, emqx_mt_ccache).
+
+%% Mria table (disc_copies) to store the namespace records.
+%% Value is for future use.
+-record(?NS_TAB, {
+    ns :: tns(),
+    value = [] :: term()
+}).
+
+%% Mria table to store the client records.
+%% Pid is used in the key to make sure the record is unique,
+%% so there is no need for transaction to update the record.
+-define(RECORD_KEY(Ns, ClientId, Pid), {Ns, ClientId, Pid}).
+%% 0 is less '<' than any pid
+-define(MIN_PID, 0).
+-define(MIN_RECORD_KEY(Ns), ?RECORD_KEY(Ns, ?MIN_CLIENTID, ?MIN_PID)).
+-record(?RECORD_TAB, {
+    key :: ?RECORD_KEY(tns(), clientid(), pid()),
+    node :: node(),
+    extra :: []
+}).
+
+%% Mria table to store the number of clients in each namespace.
+%% The key has node in it to make sure the counter is unique,
+%% so there is no need for transaction to update the counter.
+-define(COUNTER_KEY(Ns, Node), {Ns, Node}).
+-record(?COUNTER_TAB, {
+    key :: ?COUNTER_KEY(tns(), node()),
+    count :: non_neg_integer()
+}).
+
+%% ETS table to keep track of the session pid for each client.
+-define(MONITOR(Pid, Key), {Pid, Key}).
+-define(CCACHE(Ns, Ts, Cnt), {Ns, Ts, Cnt}).
+-define(CCACHE_VALID_MS, 5000).
+
+-define(LOCK(Node), {emqx_mt_clear_node_lock, Node}).
+
+-type tns() :: emqx_mt:tns().
+-type clientid() :: emqx_types:clientid().
+
+create_tables() ->
+    ok = mria:create_table(?NS_TAB, [
+        {type, ordered_set},
+        {rlog_shard, ?EMQX_MT_SHARD},
+        {storage, disc_copies},
+        {attributes, record_info(fields, ?NS_TAB)}
+    ]),
+    ok = mria:create_table(?RECORD_TAB, [
+        {type, ordered_set},
+        {rlog_shard, ?EMQX_MT_SHARD},
+        {storage, ram_copies},
+        {attributes, record_info(fields, ?RECORD_TAB)}
+    ]),
+    ok = mria:create_table(?COUNTER_TAB, [
+        {type, set},
+        {rlog_shard, ?EMQX_MT_SHARD},
+        {storage, ram_copies},
+        {attributes, record_info(fields, ?COUNTER_TAB)}
+    ]),
+    _ = ets:new(?MONITOR_TAB, [
+        set,
+        named_table,
+        public
+    ]),
+    _ = ets:new(?CCACHE_TAB, [
+        set,
+        named_table,
+        public
+    ]),
+    ok.
+
+%% @doc List namespaces.
+%% The second argument is the last namespace from the previous page.
+%% The third argument is the number of namespaces to return.
+-spec list_ns(tns(), pos_integer()) -> [tns()].
+list_ns(LastNs, Limit) ->
+    Ms = ets:fun2ms(fun(#?NS_TAB{ns = Ns}) when Ns > LastNs -> Ns end),
```

Comment:
Nit: the resulting `Ms` probably won't be able to make use of key ordering, because `when` clause goes to the condition part of the `Ms`.
But if the size of the table is supposed to be small, this may be irrelevant.

Reply:
the number of ns can potentially be very large.
because it can be used to index username to clientid
will fix